File: modules/pddl12.cpp
//
// Created by adamgreen on 24/01/19.
//

#include "pddl12.h"

#include <boost/process.hpp>
#include <boost/timer/timer.hpp>

#include "../utilities/functions.cpp"
#include "../utilities/Executor.h"
#include "../utilities/Ellipsis.h"

namespace bp = boost::process;
namespace bt = boost::timer;


bool pddl12::runCritical(TestResults &results) {
    return testStrips(results);
}

void pddl12::runNonCritical(TestResults &results) {
    testTyping(results);
    testDisjunctivePreconditions(results);
    testEquality(results);
    testExistentialPreconditions(results);
    testUniversalPreconditions(results);
    testConditionalEffects(results);
    testDomainAxioms(results);
    testSubgoalsThroughAxioms(results);
    testSafetyConstraints(results);
    testExpressionEvaluation(results);
    testOpenWorld(results);
    checkQuantifiedPreconditionsSupport(results);
    checkADLSupport(results);
    checkUCPOPSupport(results);

}

bool pddl12::testStrips(TestResults & results) {
    return conductTest("strips", "STRIPS", TestResults::pddl12Strips, results, 1);
}

void pddl12::testTyping(TestResults & results) {
    conductTest("typing", "Typing", TestResults::test::pddl12Typing, results, 1);
}

void pddl12::testDisjunctivePreconditions(TestResults & results) {
    conductTest("disjunctive-preconditions", "Disjunctive Preconditions", TestResults::test::pddl12DisjunctivePreconditions, results, 1);
}

void pddl12::testEquality(TestResults & results) {
    conductTest("equality", "Equality", TestResults::test::pddl12Equality, results, 1);
}

void pddl12::testExistentialPreconditions(TestResults & results) {
    conductTest("existential-preconditions", "Existential Preconditions", TestResults::pddl12ExistentialPreconditions, results, 1);
}

void pddl12::testUniversalPreconditions(TestResults & results) {
    conductTest("universal-preconditions", "Universal Preconditions", TestResults::pddl12UniversalPreconditions, results, 1);
}

void pddl12::testConditionalEffects(TestResults & results) {
    conductTest("conditional-effects", "Conditional Effects", TestResults::pddl12ConditionalEffects, results, 2);
}

void pddl12::testDomainAxioms(TestResults &results) {
    conductTest("domain-axioms", "Domain Axioms", TestResults::pddl12DomainAxioms, results, 2);
}

void pddl12::testSubgoalsThroughAxioms(TestResults &results) {
    conductTest("subgoals-through-axioms", "Subgoals Through Axioms", TestResults::pddl12SubgoalsThroughAxioms, results, 2);
}

void pddl12::testSafetyConstraints(TestResults &results) {
    conductTest("safety-constraints", "Safety Constraints", TestResults::pddl12SafetyConstraints, results, 3);
}

void pddl12::testExpressionEvaluation(TestResults &results) {
    conductTest("expression-evaluation", "Expression Evaluation", TestResults::pddl12ExpressionEvaluation, results, 0);
}

void pddl12::testOpenWorld(TestResults &results) {
    conductTest("open-world", "Open World", TestResults::pddl12OpenWorld, results, 1);
}

void pddl12::checkQuantifiedPreconditionsSupport(TestResults &results) {
    int desiredCode = results.getPassCode();
    std::cout << std::endl << "Testing super Requirement - Quantified Preconditions";

    bool passed = true;
    bool cannotConductFullTest = false;
    int exitCode = desiredCode;
    {
        std::cout << std::endl << "\t- Existential Preconditions supported: ";
        auto cond = results.getTestResult(TestResults::pddl12ExistentialPreconditions);
        if (cond.first) {
            if (cond.second == desiredCode) {
                std::cout << "TRUE";
            } else {
                std::cout << "FALSE";
                exitCode = cond.second;
                passed = false;
            }
        } else {
            std::cout << "PRE-REQUISITE TEST NOT COMPLETED";
            cannotConductFullTest = true;
            passed = false;
        }
    }

    {
        std::cout << std::endl << "\t- Universal Preconditions supported: ";
        auto cond = results.getTestResult(TestResults::pddl12UniversalPreconditions);
        if (cond.first) {
            if (cond.second == desiredCode) {
                std::cout << "TRUE";
            } else {
                std::cout << "FALSE";
                exitCode = cond.second;
                passed = false;
            }
        } else {
            std::cout << "PRE-REQUISITE TEST NOT COMPLETED";
            cannotConductFullTest = true;
            passed = false;
        }
    }

    if (passed) {
        std::cout << std::endl << "PASSED: Planner supports quantified preconditions (super requirement)" << std::endl << std::endl;
        results.addTestResult(TestResults::pddl12QuantifiedPreconditions, desiredCode, TestResults::passed);
    } else {
        if (cannotConductFullTest) {
            std::cout << std::endl << "INCOMPLETE: Eviscerator could not confirm if Quantified Preconditions are supported because some pre-req tests have not been finished";
        } else {
            std::cout << std::endl << "FAILED: Quantified Preconditions are not supported";
            results.addTestResult(TestResults::pddl12QuantifiedPreconditions, exitCode, TestResults::failed);
        }
    }
}

void pddl12::checkADLSupport(TestResults &results) {
    int desiredCode = results.getTestResult(TestResults::pddl12Strips).second;
    std::cout << std::endl << "Testing super Requirement - ADL";

    bool passed = true;
    bool cannotConductFullTest = false;
    int exitCode = desiredCode;
    {
        std::cout << std::endl << "\t- STRIPS supported: ";
        auto cond = results.getTestResult(TestResults::pddl12Strips);
        if (cond.first) {
            if (cond.second == desiredCode) {
                std::cout << "TRUE";
            } else {
                std::cout << "FALSE";
                exitCode = cond.second;
                passed = false;
            }
        } else {
            std::cout << "PRE-REQUISITE TEST NOT COMPLETED";
            cannotConductFullTest = true;
            passed = false;
        }
    }
    {
        std::cout << std::endl << "\t- Typing supported: ";
        auto cond = results.getTestResult(TestResults::pddl12QuantifiedPreconditions);
        if (cond.first) {
            if (cond.second == desiredCode) {
                std::cout << "TRUE";
            } else {
                std::cout << "FALSE";
                exitCode = cond.second;
                passed = false;
            }
        } else {
            std::cout << "PRE-REQUISITE TEST NOT COMPLETED";
            cannotConductFullTest = true;
            passed = false;
        }
    }
    {
        std::cout << std::endl << "\t- Disjunctive Preconditions supported: ";
        auto cond = results.getTestResult(TestResults::pddl12DisjunctivePreconditions);
        if (cond.first) {
            if (cond.second == desiredCode) {
                std::cout << "TRUE";
            } else {
                std::cout << "FALSE";
                exitCode = cond.second;
                passed = false;
            }
        } else {
            std::cout << "PRE-REQUISITE TEST NOT COMPLETED";
            cannotConductFullTest = true;
            passed = false;
        }
    }
    {
        std::cout << std::endl << "\t- Equality supported: ";
        auto cond = results.getTestResult(TestResults::pddl12Equality);
        if (cond.first) {
            if (cond.second == desiredCode) {
                std::cout << "TRUE";
            } else {
                std::cout << "FALSE";
                exitCode = cond.second;
                passed = false;
            }
        } else {
            std::cout << "PRE-REQUISITE TEST NOT COMPLETED";
            cannotConductFullTest = true;
            passed = false;
        }
    }
    {
        std::cout << std::endl << "\t- Quantified Preconditions supported: ";
        auto cond = results.getTestResult(TestResults::pddl12QuantifiedPreconditions);
        if (cond.first) {
            if (cond.second == desiredCode) {
                std::cout << "TRUE";
            } else {
                std::cout << "FALSE";
                exitCode = cond.second;
                passed = false;
            }
        } else {
            std::cout << "PRE-REQUISITE TEST NOT COMPLETED";
            cannotConductFullTest = true;
            passed = false;
        }
    }
    {
        std::cout << std::endl << "\t- Conditional Effects supported: ";
        auto cond = results.getTestResult(TestResults::pddl12ConditionalEffects);
        if (cond.first) {
            if (cond.second == desiredCode) {
                std::cout << "TRUE";
            } else {
                std::cout << "FALSE";
                exitCode = cond.second;
                passed = false;
            }
        } else {
            std::cout << "PRE-REQUISITE TEST NOT COMPLETED";
            cannotConductFullTest = true;
            passed = false;
        }
    }

    if (passed) {
        std::cout << std::endl << "PASSED: Planner supports ADL (super requirement)" << std::endl << std::endl;
        results.addTestResult(TestResults::pddl12ADL, desiredCode, TestResults::passed);
    } else {
        if (cannotConductFullTest) {
            std::cout << std::endl << "INCOMPLETE: Eviscerator could not confirm if ADL is supported because some pre-req tests have not been finished";
        } else {
            std::cout << std::endl << "FAILED: ADL is not supported";
            results.addTestResult(TestResults::pddl12ADL, exitCode, TestResults::failed);
        }
    }
}

void pddl12::checkUCPOPSupport(TestResults &results) {
    int desiredCode = results.getTestResult(TestResults::pddl12Strips).second;
    std::cout << std::endl << "Testing super Requirement - UCPOP";

    bool passed = true;
    bool cannotConductFullTest = false;
    int exitCode = desiredCode;
    {
        std::cout << std::endl << "\t- ADL supported: ";
        auto cond = results.getTestResult(TestResults::pddl12ADL);
        if (cond.first) {
            if (cond.second == desiredCode) {
                std::cout << "TRUE";
            } else {
                std::cout << "FALSE";
                exitCode = cond.second;
                passed = false;
            }
        } else {
            std::cout << "PRE-REQUISITE TEST NOT COMPLETED";
            cannotConductFullTest = true;
            passed = false;
        }
    }
    {
        std::cout << std::endl << "\t- Domain Axioms supported: ";
        auto cond = results.getTestResult(TestResults::pddl12DomainAxioms);
        if (cond.first) {
            if (cond.second == desiredCode) {
                std::cout << "TRUE";
            } else {
                std::cout << "FALSE";
                exitCode = cond.second;
                passed = false;
            }
        } else {
            std::cout << "PRE-REQUISITE TEST NOT COMPLETED";
            cannotConductFullTest = true;
            passed = false;
        }
    }
    {
        std::cout << std::endl << "\t- Safety Constraints supported: ";
        auto cond = results.getTestResult(TestResults::pddl12SafetyConstraints);
        if (cond.first) {
            if (cond.second == desiredCode) {
                std::cout << "TRUE";
            } else {
                std::cout << "FALSE";
                exitCode = cond.second;
                passed = false;
            }
        } else {
            std::cout << "PRE-REQUISITE TEST NOT COMPLETED";
            cannotConductFullTest = true;
            passed = false;
        }
    }
    if (passed) {
        std::cout << std::endl << "PASSED: Planner supports UCPOP (super requirement)" << std::endl << std::endl;
        results.addTestResult(TestResults::pddl12UCPOP, desiredCode, TestResults::passed);
    } else {
        if (cannotConductFullTest) {
            std::cout << std::endl << "INCOMPLETE: Eviscerator could not confirm if UCPOP is supported because some pre-req tests have not been finished";
        } else {
            std::cout << std::endl << "FAILED: UCPOP is not supported";
            results.addTestResult(TestResults::pddl12UCPOP, exitCode, TestResults::failed);
        }
    }
}

File: src/pddl12.rs
use crate::core::*;
use crate::execution::{Executor};
use crate::core::TestResult::{Passed, Failed};

pub fn test_critical_features(exe : &Executor, results: EvisceratorResults) -> EvisceratorResults {
    println!("\t\t--[Checking Planner can plan a STRIPS Problem]--\t\t");
    let r = exe.execute("/pddl/pddl12/domain-strips.pddl", "/pddl/pddl12/problem-strips.pddl");
    halt_and_catch_fire_if(r.exit_code < 0);
    if r.success && r.plan.len() > 0 {
        println!("PASSED: The given planner can plan a strips problem\n");
        EvisceratorResults {
            success_exit_code: r.exit_code,
            pddl12: PDDL12Tests {
                strips: Passed,
                ..results.pddl12
            },
            ..results
        }
    } else {
        println!("CRITICAL FAIL: The given planner cannot plan a strips problem\n");
        EvisceratorResults {
            pddl12: PDDL12Tests {
                strips: Failed,
                ..results.pddl12
            },
            ..results
        }
    }

}

pub fn attempt_recovery_from_strips_failure(exe : &Executor, results: EvisceratorResults) -> EvisceratorResults {
    println!("\t\t--[Checking Planner can plan a STRIPS + Typing Problem]--\t\t");
    let r = exe.execute("/pddl/pddl12/domain-typing.pddl", "/pddl/pddl12/problem-typing.pddl");
    halt_and_catch_fire_if(r.exit_code < 0);
    if r.success && r.plan.len() > 0 {
        println!("PASSED: The given planner can plan a STRIPS + Typing problem\n");
        EvisceratorResults {
            success_exit_code: r.exit_code,
            typing_dependent: true,
            pddl12: PDDL12Tests {
                strips: Passed,
                typing: Passed,
                ..results.pddl12
            },
            ..results
        }
    } else {
        println!("CRITICAL FAIL: The given planner cannot plan a STRIPS + Typing problem\n");
        EvisceratorResults {
            pddl12: PDDL12Tests {
                strips: Failed,
                ..results.pddl12
            },
            ..results
        }
    }
}

pub fn test_typing(exe: &Executor, results: EvisceratorResults) -> EvisceratorResults {
    println!("\t\t--[Checking Planner can plan a Typing Problem]--\t\t");
    let r = exe.execute("/pddl/pddl12/domain-typing.pddl", "/pddl/pddl12/problem-typing.pddl");
    halt_and_catch_fire_if(r.exit_code < 0);
    if r.success && r.plan.len() > 0 {
        println!("PASSED: The given planner can plan a Typing problem\n");
        EvisceratorResults {
            pddl12: PDDL12Tests {
                typing: Passed,
                ..results.pddl12
            },
            ..results
        }
    } else {
        println!("FAIL: The given planner cannot plan a Typing problem\n");
        EvisceratorResults {
            pddl12: PDDL12Tests {
                typing: Failed,
                ..results.pddl12
            },
            ..results
        }
    }
}

pub fn test_disjunctive_preconditions(exe: &Executor, results: EvisceratorResults) -> EvisceratorResults {
    let full_name = "Disjunctive Preconditions";
    println!("\t\t--[Checking Planner can plan a {} Problem]--\t\t", full_name);
    let r;
    match results.typing_dependent {
        true => r = exe.execute(
            "/pddl/pddl12/domain-disjunctive-preconditions-td.pddl",
            "/pddl/pddl12/problem-disjunctive-preconditions-td.pddl"),
        false => r = exe.execute(
            "/pddl/pddl12/domain-disjunctive-preconditions.pddl",
            "/pddl/pddl12/problem-disjunctive-preconditions.pddl")
    }
    halt_and_catch_fire_if(r.exit_code < 0);
    if r.success && r.plan.len() > 0 {
        println!("PASSED: The given planner can plan a {} problem\n", full_name);
        EvisceratorResults {
            pddl12: PDDL12Tests {
                disjunctive_preconditions: Passed,
                ..results.pddl12
            },
            ..results
        }
    } else {
        println!("FAIL: The given planner cannot plan a {} problem\n", full_name);
        EvisceratorResults {
            pddl12: PDDL12Tests {
                disjunctive_preconditions: Failed,
                ..results.pddl12
            },
            ..results
        }
    }
}


pub fn test_equality(exe: &Executor, results: EvisceratorResults) -> EvisceratorResults {
    let full_name = "Equality";
    println!("\t\t--[Checking Planner can plan a {} Problem]--\t\t", full_name);
    let r;
    match results.typing_dependent {
        true => r = exe.execute(
            "/pddl/pddl12/domain-equality-td.pddl",
            "/pddl/pddl12/problem-equality-td.pddl"),
        false => r = exe.execute(
            "/pddl/pddl12/domain-equality.pddl",
            "/pddl/pddl12/problem-equality.pddl")
    }
    halt_and_catch_fire_if(r.exit_code < 0);
    if r.success && r.plan.len() > 0 {
        println!("PASSED: The given planner can plan a {} problem\n", full_name);
        EvisceratorResults {
            pddl12: PDDL12Tests {
                equality: Passed,
                ..results.pddl12
            },
            ..results
        }
    } else {
        println!("FAIL: The given planner cannot plan a {} problem\n", full_name);
        EvisceratorResults {
            pddl12: PDDL12Tests {
                equality: Failed,
                ..results.pddl12
            },
            ..results
        }
    }
}

pub fn test_existential_preconditions(exe: &Executor, results: EvisceratorResults) -> EvisceratorResults {
    let full_name = "Existential Preconditions";
    println!("\t\t--[Checking Planner can plan a {} Problem]--\t\t", full_name);
    let r;
    match results.typing_dependent {
        true => r = exe.execute(
            "/pddl/pddl12/domain-existential-preconditions-td.pddl",
            "/pddl/pddl12/problem-existential-preconditions-td.pddl"),
        false => r = exe.execute(
            "/pddl/pddl12/domain-existential-preconditions.pddl",
            "/pddl/pddl12/problem-existential-preconditions.pddl")
    }
    halt_and_catch_fire_if(r.exit_code < 0);
    if r.success && r.plan.len() > 0 {
        println!("PASSED: The given planner can plan a {} problem\n", full_name);
        EvisceratorResults {
            pddl12: PDDL12Tests {
                existential_preconditions: Passed,
                ..results.pddl12
            },
            ..results
        }
    } else {
        println!("FAIL: The given planner cannot plan a {} problem\n", full_name);
        EvisceratorResults {
            pddl12: PDDL12Tests {
                existential_preconditions: Failed,
                ..results.pddl12
            },
            ..results
        }
    }
}

pub fn test_universal_preconditions(exe: &Executor, results: EvisceratorResults) -> EvisceratorResults {
    let full_name = "Universal Preconditions";
    println!("\t\t--[Checking Planner can plan a {} Problem]--\t\t", full_name);
    let r;
    match results.typing_dependent {
        true => r = exe.execute(
            "/pddl/pddl12/domain-universal-preconditions-td.pddl",
            "/pddl/pddl12/problem-universal-preconditions-td.pddl"),
        false => r = exe.execute(
            "/pddl/pddl12/domain-universal-preconditions.pddl",
            "/pddl/pddl12/problem-universal-preconditions.pddl")
    }
    halt_and_catch_fire_if(r.exit_code < 0);
    if r.success && r.plan.len() > 0 {
        println!("PASSED: The given planner can plan a {} problem\n", full_name);
        EvisceratorResults {
            pddl12: PDDL12Tests {
                universal_preconditions: Passed,
                ..results.pddl12
            },
            ..results
        }
    } else {
        println!("FAIL: The given planner cannot plan a {} problem\n", full_name);
        EvisceratorResults {
            pddl12: PDDL12Tests {
                universal_preconditions: Failed,
                ..results.pddl12
            },
            ..results
        }
    }
}

pub fn test_conditional_effects(exe: &Executor, results: EvisceratorResults) -> EvisceratorResults {
    let full_name = "Conditional Effects";
    println!("\t\t--[Checking Planner can plan a {} Problem]--\t\t", full_name);
    let r;
    match results.typing_dependent {
        true => r = exe.execute(
            "/pddl/pddl12/domain-conditional-effects-td.pddl",
            "/pddl/pddl12/problem-conditional-effects-td.pddl"),
        false => r = exe.execute(
            "/pddl/pddl12/domain-conditional-effects.pddl",
            "/pddl/pddl12/problem-conditional-effects.pddl")
    }
    halt_and_catch_fire_if(r.exit_code < 0);
    if r.success && r.plan.len() > 0 {
        println!("PASSED: The given planner can plan a {} problem\n", full_name);
        EvisceratorResults {
            pddl12: PDDL12Tests {
                conditional_effects: Passed,
                ..results.pddl12
            },
            ..results
        }
    } else {
        println!("FAIL: The given planner cannot plan a {} problem\n", full_name);
        EvisceratorResults {
            pddl12: PDDL12Tests {
                conditional_effects: Failed,
                ..results.pddl12
            },
            ..results
        }
    }
}

pub fn test_domain_axioms(exe: &Executor, results: EvisceratorResults) -> EvisceratorResults {
    let full_name = "Domain Axioms";
    println!("\t\t--[Checking Planner can plan a {} Problem]--\t\t", full_name);
    let r;
    match results.typing_dependent {
        true => r = exe.execute(
            "/pddl/pddl12/domain-domain-axioms-td.pddl",
            "/pddl/pddl12/problem-domain-axioms-td.pddl"),
        false => r = exe.execute(
            "/pddl/pddl12/domain-domain-axioms.pddl",
            "/pddl/pddl12/problem-domain-axioms.pddl")
    }
    halt_and_catch_fire_if(r.exit_code < 0);
    if r.success && r.plan.len() > 0 {
        println!("PASSED: The given planner can plan a {} problem\n", full_name);
        EvisceratorResults {
            pddl12: PDDL12Tests {
                domain_axioms: Passed,
                ..results.pddl12
            },
            ..results
        }
    } else {
        println!("FAIL: The given planner cannot plan a {} problem\n", full_name);
        EvisceratorResults {
            pddl12: PDDL12Tests {
                domain_axioms: Failed,
                ..results.pddl12
            },
            ..results
        }
    }
}

pub fn test_subgoals_through_axioms(exe: &Executor, results: EvisceratorResults) -> EvisceratorResults {
    let full_name = "Subgoals Through Axioms";
    println!("\t\t--[Checking Planner can plan a {} Problem]--\t\t", full_name);
    let r;
    match results.typing_dependent {
        true => r = exe.execute(
            "/pddl/pddl12/domain-subgoals-through-axioms-td.pddl",
            "/pddl/pddl12/problem-subgoals-through-axioms-td.pddl"),
        false => r = exe.execute(
            "/pddl/pddl12/domain-subgoals-through-axioms.pddl",
            "/pddl/pddl12/problem-subgoals-through-axioms.pddl")
    }
    halt_and_catch_fire_if(r.exit_code < 0);
    if r.success && r.plan.len() > 0 {
        println!("PASSED: The given planner can plan a {} problem\n", full_name);
        EvisceratorResults {
            pddl12: PDDL12Tests {
                subgoals_through_axioms: Passed,
                ..results.pddl12
            },
            ..results
        }
    } else {
        println!("FAIL: The given planner cannot plan a {} problem\n", full_name);
        EvisceratorResults {
            pddl12: PDDL12Tests {
                subgoals_through_axioms: Failed,
                ..results.pddl12
            },
            ..results
        }
    }
}

pub fn test_safety_constraints(exe: &Executor, results: EvisceratorResults) -> EvisceratorResults {
    let full_name = "Safety Constraints";
    println!("\t\t--[Checking Planner can plan a {} Problem]--\t\t", full_name);
    let r;
    match results.typing_dependent {
        true => r = exe.execute(
            "/pddl/pddl12/domain-safety-constraints-td.pddl",
            "/pddl/pddl12/problem-safety-constraints-td.pddl"),
        false => r = exe.execute(
            "/pddl/pddl12/domain-safety-constraints.pddl",
            "/pddl/pddl12/problem-safety-constraints.pddl")
    }
    halt_and_catch_fire_if(r.exit_code < 0);
    if r.success && r.plan.len() > 0 {
        println!("PASSED: The given planner can plan a {} problem\n", full_name);
        EvisceratorResults {
            pddl12: PDDL12Tests {
                safety_constraints: Passed,
                ..results.pddl12
            },
            ..results
        }
    } else {
        println!("FAIL: The given planner cannot plan a {} problem\n", full_name);
        EvisceratorResults {
            pddl12: PDDL12Tests {
                safety_constraints: Failed,
                ..results.pddl12
            },
            ..results
        }
    }
}

pub fn test_open_world(exe: &Executor, results: EvisceratorResults) -> EvisceratorResults {
    let full_name = "Open World";
    println!("\t\t--[Checking Planner can plan a {} Problem]--\t\t", full_name);
    let r;
    match results.typing_dependent {
        true => r = exe.execute(
            "/pddl/pddl12/domain-open-world-td.pddl",
            "/pddl/pddl12/problem-open-world-td.pddl"),
        false => r = exe.execute(
            "/pddl/pddl12/domain-open-world.pddl",
            "/pddl/pddl12/problem-open-world.pddl")
    }
    halt_and_catch_fire_if(r.exit_code < 0);
    if r.success && r.plan.len() > 0 {
        println!("PASSED: The given planner can plan a {} problem\n", full_name);
        EvisceratorResults {
            pddl12: PDDL12Tests {
                open_world: Passed,
                ..results.pddl12
            },
            ..results
        }
    } else {
        println!("FAIL: The given planner cannot plan a {} problem\n", full_name);
        EvisceratorResults {
            pddl12: PDDL12Tests {
                open_world: Failed,
                ..results.pddl12
            },
            ..results
        }
    }
}